### PR TITLE
Validate the network mapping configuration

### PR DESCRIPTION
**Pull request by votdev**
**Problem:**
The VM import does not check if the specified source and destination networks exist.

**Solution:**
During the preflight checks do a lookup on the specified networks.

- A new `apiGroups` is added to the `ClusterRole` to be able to list `network-attachment-definitions`.
- The `network-attachment-definitions` CRD has to be generated in the test env.

**Related Issue:**
https://github.com/harvester/harvester/issues/6491

**Test plan:**
# Case 1
- Create the following manifest and apply it. Make sure the destination network `default/vlan2` should not exist in Harvester.
```
apiVersion: migration.harvesterhci.io/v1beta1
kind: VirtualMachineImport
metadata:
  name: cirros-tiny
  namespace: default
spec:
  virtualMachineName: "cirros-tiny"
  networkMapping:
  - sourceNetwork: "shared"
    destinationNetwork: "default/vlan2"
  sourceCluster:
    name: devstack
    namespace: default
    kind: OpenstackSource
    apiVersion: migration.harvesterhci.io/v1beta1
```
- The log output should look like this:
```
...
time="2024-09-26T07:21:30Z" level=info msg="Reconciling migration source" kind=VmwareSource name=vcsim namespace=default
time="2024-09-26T07:21:30Z" level=info msg="Starting migration.harvesterhci.io/v1beta1, Kind=OpenstackSource controller"
time="2024-09-26T07:21:30Z" level=info msg="Reconciling migration source" kind=OpenstackSource name=devstack namespace=default
time="2024-09-26T07:21:30Z" level=info msg="Starting k8s.cni.cncf.io/v1, Kind=NetworkAttachmentDefinition controller"
time="2024-09-26T07:22:29Z" level=info msg="Running preflight checks ..." name=cirros-tiny namespace=default spec.virtualMachineName=cirros-tiny
time="2024-09-26T07:22:29Z" level=error msg="Failed to get destination network 'default/vlan2': network-attachment-definitions.k8s.cni.cncf.io \"vlan2\" not found" name=cirros-tiny namespace=default spec.sourcecluster.kind=OpenstackSource
time="2024-09-26T07:22:29Z" level=error msg="error syncing 'default/cirros-tiny': handler virtualmachine-import-job-change: network-attachment-definitions.k8s.cni.cncf.io \"vlan2\" not found, requeuing"
...
```

# Case 2
- Create the following manifest and apply it. Make sure the source network `foo` should not exist in OpenStack.
```
apiVersion: migration.harvesterhci.io/v1beta1
kind: VirtualMachineImport
metadata:
  name: cirros-tiny
  namespace: default
spec:
  virtualMachineName: "cirros-tiny"
  networkMapping:
  - sourceNetwork: "foo"
    destinationNetwork: "default/vlan1"
  sourceCluster:
    name: devstack
    namespace: default
    kind: OpenstackSource
    apiVersion: migration.harvesterhci.io/v1beta1
```
- The log output should look like this:
```
...
time="2024-09-26T07:31:29Z" level=info msg="Reconciling migration source" kind=OpenstackSource name=devstack namespace=default
time="2024-09-26T07:31:29Z" level=info msg="Starting migration.harvesterhci.io/v1beta1, Kind=VirtualMachineImport controller"
time="2024-09-26T07:31:29Z" level=info msg="Starting storage.k8s.io/v1, Kind=StorageClass controller"
time="2024-09-26T07:31:29Z" level=info msg="Starting migration.harvesterhci.io/v1beta1, Kind=VmwareSource controller"
time="2024-09-26T07:31:29Z" level=info msg="Reconciling migration source" kind=VmwareSource name=vcsim namespace=default
time="2024-09-26T07:31:35Z" level=info msg="Running preflight checks ..." name=cirros-tiny namespace=default spec.virtualMachineName=cirros-tiny
time="2024-09-26T07:31:36Z" level=error msg="Failed to perform source cluster specific preflight checks: error getting source network 'foo': Resource not found: [GET http://xxx/compute/v2.1/networks/foo], error message: {\"message\": \"The resource could not be found.<br /><br />\\n\\n\\n\", \"code\": \"404 Not Found\", \"title\": \"Not Found\"}" name=cirros-tiny namespace=default spec.sourcecluster.kind=OpenstackSource
time="2024-09-26T07:31:36Z" level=error msg="error syncing 'default/cirros-tiny': handler virtualmachine-import-job-change: error getting source network 'foo': Resource not found: [GET http://xxx/compute/v2.1/networks/foo], error message: {\"message\": \"The resource could not be found.<br /><br />\\n\\n\\n\", \"code\": \"404 Not Found\", \"title\": \"Not Found\"}, requeuing"
...
```

# Case 3
- Create the following manifest and apply it. Make sure the source network `VM Network 2` should not exist in Vmware.
```
apiVersion: migration.harvesterhci.io/v1beta1
kind: VirtualMachineImport
metadata:
  name: mr-leap1
  namespace: default
spec: 
  virtualMachineName: "mr-leap1"
  networkMapping:
  - sourceNetwork: "VM Network 2"
    destinationNetwork: "default/vlan1"
  sourceCluster: 
    name: vcsim
    namespace: default
    kind: VmwareSource
    apiVersion: migration.harvesterhci.io/v1beta1
```
- The log output should look like this:
```
...
time="2024-09-26T07:36:33Z" level=info msg="Reconciling migration source" kind=OpenstackSource name=devstack namespace=default
time="2024-09-26T07:36:33Z" level=info msg="Starting migration.harvesterhci.io/v1beta1, Kind=VirtualMachineImport controller"
time="2024-09-26T07:36:33Z" level=info msg="Starting migration.harvesterhci.io/v1beta1, Kind=VmwareSource controller"
time="2024-09-26T07:36:33Z" level=info msg="Starting storage.k8s.io/v1, Kind=StorageClass controller"
time="2024-09-26T07:36:33Z" level=info msg="Reconciling migration source" kind=VmwareSource name=vcsim namespace=default
time="2024-09-26T07:36:43Z" level=info msg="Running preflight checks ..." name=mr-leap1 namespace=default spec.virtualMachineName=mr-leap1
time="2024-09-26T07:36:45Z" level=error msg="Failed to perform source cluster specific preflight checks: error getting source network 'VM Network 2': network '/Datacenter/network/VM Network 2' not found" name=mr-leap1 namespace=default spec.sourcecluster.kind=VmwareSource
time="2024-09-26T07:36:45Z" level=error msg="error syncing 'default/mr-leap1': handler virtualmachine-import-job-change: error getting source network 'VM Network 2': network '/Datacenter/network/VM Network 2' not found, requeuing"
...
```

### CODE REVIEW

File: pkg/controllers/migration/virtualmachine.go [hunk 1]
```diff
@@ -237,6 +243,55 @@ func (h *virtualMachineHandler) preFlightChecks(vm *migration.VirtualMachineImpo
 		return fmt.Errorf("source network %s appears multiple times in vm spec", network.SourceNetwork)
 	}
 
+	// Validate the destination network configuration.
+	for _, nm := range vm.Spec.Mapping {
+		// The destination network supports the following format:
+		// - <namespace>
```
[the remark this answers is not in the document]
Comment (votdev) [reply pkg/controllers/migration/virtualmachine.go:249]: Good spotting. Fixed

File: tests/setup/harvester_mock.go [hunk 2]
```diff
@@ -37,10 +38,23 @@ func InstallCRD(ctx context.Context, cfg *rest.Config) error {
...
+		return nil, err
+	}
+	return append(kubeVirtCrds, k8sCniCncfIoCrds...), nil
+}
+
 type kubeVirtCRDGenerator func() (*extv1.CustomResourceDefinition, error)
 
-// GenerateKubeVirtCRD's will generate kubevirt CRDs
-func GenerateKubeVirtCRD() ([]*extv1.CustomResourceDefinition, error) {
+// GenerateKubeVirtCRD will generate kubevirt CRDs
```
[the remark this answers is not in the document]
Comment (votdev) [reply tests/setup/harvester_mock.go:56]: Fixed

**Review by starbops — APPROVED** [review]

**Comment by ibrokethecloud** [comment]
can we please also create a charts PR to include the rbac changes needed for this change https://github.com/harvester/charts/tree/master/charts/harvester-vm-import-controller

File: pkg/source/vmware/client.go [hunk 3]
```diff
@@ -123,6 +123,23 @@ func (c *Client) Verify() error {
...
 
+func (c *Client) PreFlightChecks(vm *migration.VirtualMachineImport) (err error) {
+	f := find.NewFinder(c.Client.Client, true)
+	dc := c.dc
+	if !strings.HasPrefix(c.dc, "/") {
+		dc = fmt.Sprintf("/%s", c.dc)
+	}
+	for _, nm := range vm.Spec.Mapping {
+		// The path looks like `/<datacenter>/network/<folder>/<network-name>`.
+		path := filepath.Join(dc, "/network", vm.Spec.Folder, nm.SourceNetwork)
```
Comment (ibrokethecloud) [inline pkg/source/vmware/client.go:134]: network objects exist at the dc level and are not scoped to the vm folder. I do not think we need `vm.Spec.Folder` here.

sample from one environment
```
(⎈|default:N/A)➜  ~ govc find / -type n
/Datacenter/network/VM Network
```

Comment (ibrokethecloud) [reply pkg/source/vmware/client.go:134]: rest of the changes look good

**Comment by votdev** [comment]
> can we please also create a charts PR to include the rbac changes needed for this change [harvester/charts@`master`/charts/harvester-vm-import-controller](https://github.com/harvester/charts/tree/master/charts/harvester-vm-import-controller?rgh-link-date=2024-10-07T23%3A42%3A52Z)

Done, see https://github.com/harvester/charts/pull/301.

File: pkg/source/vmware/client.go (hunk 3 above)
Comment (votdev) [reply pkg/source/vmware/client.go:134]: When i look into the code of [Finder.Network](https://github.com/vmware/govmomi/blob/5aff45ed941ee6c5648b1264b495a8108b36c999/find/finder.go#L798) there is a cluster path example and the text is mentioning `network folder`.

**Review by ibrokethecloud — APPROVED** [review]
lgtm. thanks.

**Pull request merged**